src/core/mac/mac_frame.cpp

```diff
@@ -145,6 +145,142 @@ ThreadError Frame::InitMacHeader(uint16_t aFcf, uint8_t aSecurityControl)
     return kThreadError_None;
 }
 
+ThreadError Frame::ValidatePsdu(void)
+{
+    ThreadError error = kThreadError_Parse;
+    uint8_t offset = 0;
+    uint16_t fcf;
+    uint8_t footerLength = kFcsSize;
+
+    VerifyOrExit(GetPsduLength() > kFcfSize + kDsnSize, ;);
+
+    fcf = static_cast<uint16_t>((GetPsdu()[1] << 8) | GetPsdu()[0]);
+
+    offset += kFcfSize + kDsnSize;
+
+    // Destinatinon PAN + Address
+    switch (fcf & Frame::kFcfDstAddrMask)
+    {
+    case Frame::kFcfDstAddrNone:
+        break;
+
+    case Frame::kFcfDstAddrShort:
+        VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(PanId) + sizeof(ShortAddress), ;);
```

Comment:
I like the code to check to ensure mac header/footer parse ok and that the frame contains enough bytes.
Suggestion: Instead of checking if the remaining length is ok at each step of parsing the `fcf` fields, I think it may be cleaner and easier if we just calculate the header  length based on the `fcf` fields (which is already happening by keeping track of the `offset` variable) and at the end of parsing all the related `fcf` fields (like SrcMask, DstMask), we can check if the received frame's length is larger than the expected header length. Thoughts?